File: api/src/main/java/api/security/exceptions/ErrorHandler.java
package api.security.exceptions;

import com.fasterxml.jackson.databind.ObjectMapper;
import api.security.model.ErrorResponse;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.web.access.AccessDeniedHandler;
import org.springframework.stereotype.Component;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Date;

@Component
public class ErrorHandler implements AccessDeniedHandler {

    public void handle(HttpServletResponse response, Exception e) throws IOException {

        response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
        response.setContentType(MediaType.APPLICATION_JSON_VALUE);

        new ObjectMapper().writeValue(response.getWriter(),
                new ErrorResponse(HttpStatus.UNAUTHORIZED, e.getMessage(), new Date()));
    }

    @Override
    public void handle(HttpServletRequest request, HttpServletResponse response,
                       AccessDeniedException accessDeniedException) throws IOException, ServletException {
        handle(response, accessDeniedException);
    }
}

File: api/src/main/java/api/security/exceptions/ErrorHandler.kt
package api.security.exceptions

import com.fasterxml.jackson.databind.ObjectMapper
import api.security.model.ErrorResponse
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.access.AccessDeniedHandler
import org.springframework.security.web.authentication.AuthenticationFailureHandler
import org.springframework.stereotype.Component

import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
import java.io.IOException
import java.util.Date

@Component
class ErrorHandler : AccessDeniedHandler, AuthenticationFailureHandler {

    override fun onAuthenticationFailure(request: HttpServletRequest, response: HttpServletResponse, exception: AuthenticationException) {
        handle(response, exception)
    }

    @Throws(IOException::class)
    fun handle(response: HttpServletResponse, e: Exception) {

        response.status = HttpServletResponse.SC_UNAUTHORIZED
        response.contentType = MediaType.APPLICATION_JSON_VALUE

        ObjectMapper().writeValue(response.writer,
                ErrorResponse(HttpStatus.UNAUTHORIZED, e.message))
    }

    @Throws(IOException::class, ServletException::class)
    override fun handle(request: HttpServletRequest, response: HttpServletResponse,
                        accessDeniedException: AccessDeniedException) {
        handle(response, accessDeniedException)
    }
}
